File: datasets/python_source-code/Multiple-C1/unique_1.py
"""
Graph BFS - Unique Submission
A genuinely distinct approach: BFS expressed as a lazy generator
that yields nodes one at a time, rather than building a full list.
"""

from collections import deque


def bfs_generator(graph, start):
    """Yield nodes in BFS order lazily, one at a time."""
    seen = {start}
    q = deque([start])

    while q:
        node = q.popleft()
        yield node
        for nxt in graph.get(node, []):
            if nxt not in seen:
                seen.add(nxt)
                q.append(nxt)


def take_until(gen, stop_node):
    """Consume the generator until (and including) stop_node."""
    collected = []
    for node in gen:
        collected.append(node)
        if node == stop_node:
            break
    return collected


if __name__ == "__main__":
    graph = {
        "root": ["left", "right"],
        "left": ["root", "left.left"],
        "right": ["root", "right.right"],
        "left.left": ["left"],
        "right.right": ["right"],
    }

    gen = bfs_generator(graph, "root")
    print("Full lazy traversal:", list(bfs_generator(graph, "root")))
    print("Stopping early at 'right':", take_until(bfs_generator(graph, "root"), "right"))

File: server/datasets/java_source-code/Multiple-C1/unique_1.java
/*
 * Graph BFS - Unique Submission
 * Distinct approach: BFS iterator returning nodes one by one.
 */

import java.util.*;

class Solution {
    public static List<String> bfsGenerator(Map<String, List<String>> graph, String start) {
        Set<String> seen = new HashSet<>();
        Queue<String> q = new ArrayDeque<>();
        List<String> result = new ArrayList<>();

        seen.add(start);
        q.add(start);

        while (!q.isEmpty()) {
            String node = q.poll();
            result.add(node);
            List<String> neighbors = graph.getOrDefault(node, Collections.emptyList());
            for (String nxt : neighbors) {
                if (!seen.contains(nxt)) {
                    seen.add(nxt);
                    q.add(nxt);
                }
            }
        }
        return result;
    }

    public static List<String> takeUntil(List<String> traversal, String stopNode) {
        List<String> collected = new ArrayList<>();
        for (String node : traversal) {
            collected.add(node);
            if (node.equals(stopNode)) {
                break;
            }
        }
        return collected;
    }

    public static void main(String[] args) {
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("root", Arrays.asList("left", "right"));
        graph.put("left", Arrays.asList("root", "left.left"));
        graph.put("right", Arrays.asList("root", "right.right"));
        graph.put("left.left", Arrays.asList("left"));
        graph.put("right.right", Arrays.asList("right"));

        List<String> full = bfsGenerator(graph, "root");
        System.out.println("Full lazy traversal: " + full);
        System.out.println("Stopping early at 'right': " + takeUntil(full, "right"));
    }
}
